File: step_01_basics/1.1_logic_building/P022_pattern_22_the_number_pattern/solution.py
"""
Problem: Pattern 22 - The Number Pattern
Difficulty: MEDIUM | XP: 15

For n=4, print a (2n-1)x(2n-1) grid where each cell value equals
n - min(distance to top, bottom, left, right border).

Example (n=4):
4444444
4333334
4322234
4321234
4322234
4333334
4444444
"""


# ============================================================
# APPROACH 1: BRUTE FORCE -- Per-Cell Distance Computation
# Time: O(n^2)  |  Space: O(1)
# For each cell, compute min of 4 border distances explicitly.
# ============================================================
def brute_force(n: int) -> None:
    size = 2 * n - 1
    for i in range(size):
        for j in range(size):
            dist_top    = i
            dist_bottom = size - 1 - i
            dist_left   = j
            dist_right  = size - 1 - j
            val = n - min(dist_top, dist_bottom, dist_left, dist_right)
            print(val, end="")
        print()


# ============================================================
# APPROACH 2: OPTIMAL -- Same Formula, String-Per-Row
# Time: O(n^2)  |  Space: O(n)
# Identical logic but collects digits into a list per row.
# ============================================================
def optimal(n: int) -> None:
    size = 2 * n - 1
    for i in range(size):
        row = []
        for j in range(size):
            row.append(str(n - min(i, j, size - 1 - i, size - 1 - j)))
        print("".join(row))


# ============================================================
# APPROACH 3: BEST -- Full Grid Built then Single Print
# Time: O(n^2)  |  Space: O(n^2)
# Build all rows, join and print in one I/O call.
# ============================================================
def best(n: int) -> None:
    size = 2 * n - 1
    rows = []
    for i in range(size):
        row = []
        for j in range(size):
            row.append(str(n - min(i, j, size - 1 - i, size - 1 - j)))
        rows.append("".join(row))
    print("\n".join(rows))


if __name__ == "__main__":
    N = 4
    print("=== Pattern 22 - The Number Pattern ===")
    print("--- Brute Force ---")
    brute_force(N)
    print("--- Optimal ---")
    optimal(N)
    print("--- Best ---")
    best(N)

File: step_01_basics/1.1_logic_building/P022_pattern_22_the_number_pattern/Solution.java
/**
 * Problem: Pattern 22 - The Number Pattern
 * Difficulty: MEDIUM | XP: 15
 *
 * For n=4, print a (2n-1)x(2n-1) grid:
 *   4444444
 *   4333334
 *   4322234
 *   4321234
 *   4322234
 *   4333334
 *   4444444
 *
 * Cell value = n - min(distTop, distBottom, distLeft, distRight)
 *
 * @author DSA_Nova
 */
public class Solution {

    // ============================================================
    // APPROACH 1: BRUTE FORCE -- Four Explicit Distance Variables
    // Time: O(n^2)  |  Space: O(1)
    // Compute each of the 4 border distances explicitly per cell.
    // ============================================================
    static class BruteForce {
        public static void printPattern(int n) {
            int size = 2 * n - 1;
            for (int i = 0; i < size; i++) {
                for (int j = 0; j < size; j++) {
                    int distTop    = i;
                    int distBottom = size - 1 - i;
                    int distLeft   = j;
                    int distRight  = size - 1 - j;
                    int minDist = Math.min(
                        Math.min(distTop, distBottom),
                        Math.min(distLeft, distRight)
                    );
                    System.out.print(n - minDist);
                }
                System.out.println();
            }
        }
    }

    // ============================================================
    // APPROACH 2: OPTIMAL -- StringBuilder per Row
    // Time: O(n^2)  |  Space: O(n)
    // Same formula; collect digits in a StringBuilder per row.
    // ============================================================
    static class Optimal {
        public static void printPattern(int n) {
            int size = 2 * n - 1;
            for (int i = 0; i < size; i++) {
                StringBuilder row = new StringBuilder();
                for (int j = 0; j < size; j++) {
                    int minDist = Math.min(
                        Math.min(i, size - 1 - i),
                        Math.min(j, size - 1 - j)
                    );
                    row.append(n - minDist);
                }
                System.out.println(row);
            }
        }
    }

    // ============================================================
    // APPROACH 3: BEST -- Full Grid in StringBuilder, Single Print
    // Time: O(n^2)  |  Space: O(n^2)
    // Build entire grid string, print in one call.
    // ============================================================
    static class Best {
        public static void printPattern(int n) {
            int size = 2 * n - 1;
            StringBuilder grid = new StringBuilder();
            for (int i = 0; i < size; i++) {
                for (int j = 0; j < size; j++) {
                    int minDist = Math.min(
                        Math.min(i, size - 1 - i),
                        Math.min(j, size - 1 - j)
                    );
                    grid.append(n - minDist);
                }
                if (i < size - 1) grid.append("\n");
            }
            System.out.println(grid);
        }
    }

    public static void main(String[] args) {
        int n = 4;
        System.out.println("=== Pattern 22 - The Number Pattern ===");
        System.out.println("--- Brute Force ---");
        BruteForce.printPattern(n);
        System.out.println("--- Optimal ---");
        Optimal.printPattern(n);
        System.out.println("--- Best ---");
        Best.printPattern(n);
    }
}
